splice_machine/src/test/java/com/splicemachine/derby/test/framework/SpliceWatcher.java

```diff
@@ -160,21 +173,61 @@ public PreparedStatement prepareStatement(String sql) throws SQLException {
         return ps;
     }
 
+    /**
+     * Try closing connections gracefully from different threads, if that takes long it could be some queries are stuck,
+     * so abort the connections forcefully instead
+     */
     private void closeConnections() {
+        int numThreads = connections.size() + 2; // two extra for the .abort() call
+        ExecutorService executor = Executors.newFixedThreadPool(numThreads,
+                new ThreadFactoryBuilder().setDaemon(true).setNameFormat("connection-close-%d").build());
+        CompletionService<Void> completionService =
+                new ExecutorCompletionService<>(executor);
         try {
-            for (Connection connection : connections) {
-                if (connection != null && !connection.isClosed()) {
-                    try {
-                        connection.close();
-                    } catch (SQLException e) {
-                        connection.rollback();
+            connections.stream().forEach(connection -> completionService.submit(() -> {
+                try {
+                    if (connection != null && !connection.isClosed())
                         connection.close();
+                } catch (SQLException e) {
+                }
+                return null;
+            }));
+
+            int futures = 0;
+            while(futures++ < connections.size()) {
+                Future<Void> future = completionService.poll(10, TimeUnit.SECONDS);
```

Comment:
(minor comment) this would wait 10 second not for each, but 10 second every time you call it, so the first one is for 10s, the second 20s, third 30s etc.. Maybe sth like
`Future<Void> future = completionService.poll(futures  == 0 ? 10 : 1, TimeUnit.SECONDS);`

I'm not entirely sure you need a CompletionService here at all, maybe a simple
`ExecutorService.invokeAll(Callable, timeout, TimeUnit)` would work.

Reply:
I like the invokeAll approach, we need to check the futures afterwards to see if they are done or not, but I think it's cleaner.